### ISSUE TITLE
Postinstall disabled, but still runs

### ISSUE THREAD

**Issue by roncsak** [nodkz/mongodb-memory-server#225]
Hello!

Thank you for this package!

I have a build environment which uses a docker image to build. Unfortunately I cannot use auto-download, baceuse corporate proxy and http-proxy-agent fails..

However, I can download the mongod binary with wget before installing the mongodb-memory-server package.

I downloded the package, decompressed, set +x, and exported the MONGOMS_SYSTEM_BINARY to the absolute path of the binary (/builds/company/project-core/node_modules/mongodb-binary/bin/mongod)

`mongod --version` prints the following:

    db version v4.0.3
    git version: 7ea530946fa7880364d88c8d8b6026bbc9ffa48c
    OpenSSL version: OpenSSL 1.1.0k  28 May 2019
    allocator: tcmalloc
    modules: none
    build environment:
        distmod: debian92
        distarch: x86_64
        target_arch: x86_64

However the build fails with:

    error /builds/company/project-core/node_modules/mongodb-memory-server: Command failed.
    Exit code: 1
    Command: node ./postinstall.js
    Arguments: 
    Directory: /builds/company/project-core/node_modules/mongodb-memory-server
    Output:
    mongodb-memory-server: checking MongoDB binaries cache...
    events.js:186
          throw er; // Unhandled 'error' event
          ^

    Error: unable to get local issuer certificate
        at TLSSocket.onConnectSecure (_tls_wrap.js:1317:34)
        at TLSSocket.emit (events.js:209:13)
        at TLSSocket._finishInit (_tls_wrap.js:792:8)
        at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:606:12)
    Emitted 'error' event on ClientRequest instance at:
        at TLSSocket.socketErrorListener (_http_client.js:406:9)
        at TLSSocket.emit (events.js:209:13)
        at emitErrorNT (internal/streams/destroy.js:91:8)
        at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
        at processTicksAndRejections (internal/process/task_queues.js:80:21) {
      code: 'UNABLE_TO_GET_ISSUER_CERT_LOCALLY'
    }

Since MONGOMS_SYSTEM_BINARY is set, I would not expect that mongodb-memory-server wants to download in the postinstall step.

Could you help me what am I missing, please?

**Comment by AJRdev** [comment]
@roncsak Hi ! 

Did you try to set the env variable `MONGOMS_DISABLE_POSTINSTALL` to 1 as well ?

Normally it should skip the **postinstall** script and then your `MONGOMS_SYSTEM_BINARY` path should be used for the mongo binary 😃

**Comment by nodkz** [comment]
Or use core package directly without autodownload https://github.com/nodkz/mongodb-memory-server#without-auto-download

**Comment by cx1111** [comment]
I think this issue shows that the documentation should be more clear about what each variables do, the difference between environment and package, and which stages they actually affect. ie. Upon install, upon new memory server instance, etc.

Eg. one confusing thing is that even if you install `mongodb-memory-server-core`, when creating a new instance of the memory server, it will try to auto-download a mongo binary. It just doesn't download immediately post npm install.

**Comment by hasezoey** [comment]
@cx1111 @roncsak is this still an issue with the latest versions?

**Comment by roncsak** [comment]
Hello,
I bypassed the problem.

**Comment by LilSebastian5000** [comment]
> @cx1111 @roncsak is this still an issue with the latest versions?

This is still an issue. My configuration disables postinstall and sets a system binary path in package.json, but it still attempts to download mongod when building in my pipeline:

`"config": {
    "mongodbMemoryServer": {
      "version": "4.0.3",
      "debug": "1",
      "disablePostinstall": "1",
      "systemBinary": "/tmp/app/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.3/mongod"
    }
  }`

**Comment by hasezoey** [comment]
@LilSebastian5000 thanks for the response, do i understand correctly from the output given:
- that it still tries to donwload a binary
- and seemingly does not try to use the package configs' config?

and just to confirm, you are running version 6.2.1 (package)?

PS: i havent looked into this part of the code yet, so it might take a while

**Comment by hasezoey** [comment]
@LilSebastian5000 could you provide debug output?

**Comment by Bennyelg** [comment]
+1 stuck here

**Comment by Bennyelg** [comment]
The issue was solved after 2 days by force download inside the docker container the Mongo, install it and pass to the package.json section config with the path to the  systemBinary which is /usr/bin/mongd

**Comment by hasezoey** [comment]
maybe #187 is related?

**Comment by hasezoey** [comment]
Outsourced the Proposal that "systemBinary" disables postinstall by default into #370

**Comment by hasezoey** [comment]
if someone should experience this error, please provide & make sure of the following:
- how was `disablePostinstall` configured: Environment Variable or package.json?
- if package.json, is it configured in the top-most package.json?
- what package was used from this repository?
- please provide debug output ([set environment variable `MONGOMS_DEBUG=1` before running the script](https://github.com/nodkz/mongodb-memory-server#enable-debug-mode))
- on what system-platform do you experience this error (linux, osx, win32)?

**Comment by hasezoey** [comment]
this issue might be fixed by #405 
and issue #378 and #174 might be related

**Comment by sinashk78** [comment]
Hi.
I created a project and had no issues what so ever with mongo-memory-server but when my friend tried to run the test on his machine all the test timed out and mongo-memory-server started downloading. It would download every time we ran the tests.
this is what we tried and didn’t work:
* setting MONGOMS_DISABLE_POSTINSTALL to 1 before creating an instance oo the MemoryServer and it didnt work
* passing the specific version of the ```new  MongoMemoryServer({  binary:  {  version:  "4.4.0"  }  });```

And then we tried adding the config below and setting the binary

```
"config": {
    "mongodbMemoryServer": {
      "version": "6.9.2",
      "debug": "1",
      "disablePostinstall": "1",
      "systemBinary": "mongod path"
    }
  }
```
and specifying the version like before ```new MongoMemoryServer({ binary: { version: "4.4.0" } });``` and everything worked!
and while we're trying to work things out on his machine everything was working on my end.
but what’s weird is that after removing the config from package.json it still worked.
and then we thought maybe if we just use ```new MongoMemoryServer()``` everything should work and it didn't!
To sum up, we passed the version when creating an instance and run the tests once with the config then removed the config and everything was fine we even changed the config like below and it still worked:
```
"config": {
    "mongodbMemoryServer": {
      "version": "6.9.2",
      "systemBinary": "mongod path"
    }
  }
```
If you think I’m giving false information please tell me to remove this so others won’t get confused by it. I’m not that experienced and just stated what happened to us and how we solved it.

**Comment by hasezoey** [comment]
@sinashk78 

> setting MONGOMS_DISABLE_POSTINSTALL to 1 before creating an instance oo the MemoryServer and it didnt work

`DISABLE_POSTINSTALL` is to disable the *npm hook `postinstall`*, not the runtime one, i think for that you search for #355

> And then we tried adding the config below and setting the binary
> "version": "6.9.2",

the `VERSION` config option is to set the *mongod version* to **download**, not the package version, for that you have the `dependencies` / `devDependencies` section

(and version of mongodb above `4.4` dosnt exist yet)

> "systemBinary": "mongod path"

in 6.9.x the `systemBinary` is checked if it exists and is an file and then used, otherwise an mongod file will be downloaded according to the other options

> and while we're trying to work things out on his machine everything was working on my end.

when you use the same config options for both systems (yours and your friends), then it is most likely that the `SYSTEM_BINARY` path just dosnt exists for your friend

**Comment by hasezoey** [comment]
until proven otherwise, i will close this as fixed by #405
